modules/private.php

```diff
@@ -0,0 +1,341 @@
+		}
+		
+		
+		// Make sure we are in the right code path, if not bail now.
+		
+		if ( ! is_admin() || ( ! $is_ajax_request && ! $is_admin_post_request ) ) {
+			return;
+		}
+
+		if ( ! Jetpack_Private::is_private_blog_user( $wpdb->blogid, $this_current_user ) ) {
+			wp_die();
+		}
+	}
+
+	/**
+	 * Prevents ajax requests on private blogs for users who don't have permissions
+	 *
+	 * @param string    $action The Ajax nonce action.
+	 * @param false|int $result The result of the nonce check.
+	 */
+	static function private_blog_ajax_nonce_check( $action, $result ) {
+		global $wpdb;
+
+		$this_current_user = wp_get_current_user();
+
+		if ( is_super_admin() ) {
+			return;
+		}
+
+		if ( 1 !== $result && 2 !== $result ) {
+			return;
+		}
+
+		if ( 'find-posts' !== $action && 'internal-linking' !== $action ) {
+			return;
+		}
+
+		// Make sure we are in the right code path, if not bail now.
+		if ( ! is_admin() || ( ! defined( 'DOING_AJAX' ) || ! DOING_AJAX ) ) {
+			return;
+		}
+
+		if ( ! Jetpack_Private::is_private_blog_user( $wpdb->blogid, $this_current_user ) ) {
+			wp_die( -1 );
+		}
+	}
+
+	/**
+	 * Disables WordPress Rest API for external requests
+	 */
+	static function disable_rest_api() {
+		global $wpdb;
+		$this_current_user = wp_get_current_user();
+
+		if ( is_user_logged_in() && ( is_super_admin() || Jetpack_Private::is_private_blog_user( $wpdb->blogid, $this_current_user ) ) ) {
+			return;
+		}
+
+		if ( defined( 'REST_REQUEST' ) && REST_REQUEST ) {
+			return new WP_Error( 'private_site', __( 'This site is private.', 'jetpack' ), 403 );
+		}
+	}
+
+	static function module_override( $modules ) {
+		$disabled_modules = array(
+			'publicize',
+			'sharedaddy',
+			'subscriptions',
+			'json-api',
+			'enhanced-distribution',
+			'google-analytics',
+			'photon',
+			'sitemaps',
+			'verification-tools',
+			'wordads',
+		);
+
+		foreach ( $disabled_modules as $module_slug ) {
+			$found = array_search( $module_slug, $modules );
+			if ( false !== $found ) {
+				unset( $modules[ $found ] );
+			}
+		}
+
+		return $modules;
+	}
+
+	/**
+	 * Show an error when the blog_public option is updated
+	 */
+	static function private_update_option_blog_public() {
+		if ( function_exists( 'add_settings_error') ) {
+			add_settings_error( 'general', 'setting_not_updated', sprintf( __( 'This setting is ignored because you <a href="%s">made your site private</a>', 'jetpack' ), admin_url( 'admin.php?page=jetpack' ) . '#/traffic?term=private' ), 'error' );
+		}
+	}
+
+	/**
+	 * Adds a message to the 'At a Glance' dashboard widget.
+	 */
+	static function add_private_dashboard_glance_items( $content ) {
+		return $content . '<br><br>' . __( 'This site is currently Private', 'jetpack' );
+	}
+
+	/**
+	 * Changes a blog to private
+	 */
+	static function make_blog_private() {
+		update_option( 'blog_public', -1 );
+	}
+
+	/**
+	 * Changes a blog to public
+	 */
+	static function make_blog_public() {
+		update_option( 'blog_public', 1 );
```

Comment:
I've taken a pass at implementing this in [874ad13](https://github.com/Automattic/jetpack/pull/10333/commits/874ad13c94d19b4a8c6c813ca93c5419102fc020).

> If we decide to use blog_public, could we just add a filter to either pre_option_blog_public or option_blog_public? That way, if the module were deactivated, the option would be left in whatever state it started.

I tried this but the filter does not apply for Sitemap module's [call](https://github.com/Automattic/jetpack/blob/master/modules/sitemaps.php#L22) to `get_option()`. Couple of things I tried:

Method 1: 

```
add_filter( 'pre_option_blog_public', array( __CLASS__, 'make_blog_private' ) );

static function make_blog_private( $blog_public ) {
    return -1;
}
```

Method 2:
```
add_action( 'init', array( __CLASS__, 'add_blog_private_filter' ) );

static function add_blog_private_filter() {
    add_filter( 'pre_option_blog_public', array( __CLASS__, 'make_blog_private' ) );
}

static function make_blog_private( $blog_public ) {
    return -1;
}


```
So, I've now settled for adding `update_option()` to `jetpack_activate_module_private` hook, which works in in [874ad13](https://github.com/Automattic/jetpack/pull/10333/commits/874ad13c94d19b4a8c6c813ca93c5419102fc020). 

> Running update_option() on every page load won't actually write to the database each time (since the option doesn't change), which is good, but it does still run a few filters that may be problematic.

I've moved the `update_option()` out of init to when the Private module is activated with the `jetpack_activate_module_private` hook. `blog_public`'s value is also being saved in a separate option, which is used to restore its original value when the Private site module is disabled.

Do you see any concerns to this method?

Reply:
I still think we can get this working without any option updates.

Method 2:
The problem here is that Jetpack's "modules" are all loaded on `plugins_loaded`, which happens before init. So the sitemaps module loads before the `pre_option` filter is added.

Method 1:
I suspect this isn't working because the sitemaps module is loaded before the private-sites module. If that is the problem, I see two potential solutions:
a. Ensure we always load private-sites first in `Jetpack::load_modules()`.
b. Rework sitemaps so that it doesn't include itself right away:

```diff
diff --git a/modules/sitemaps.php b/modules/sitemaps.php
index febf5786d..185c659b5 100644
--- a/modules/sitemaps.php
+++ b/modules/sitemaps.php
@@ -19,8 +19,12 @@ if ( ! defined( 'ABSPATH' ) ) {
 	exit;
 }
 
-if ( '1' == get_option( 'blog_public' ) ) { // loose comparison okay.
-	include_once 'sitemaps/sitemaps.php';
+add_action( 'plugins_loaded', 'jetpack_sitemap_load', 105 ); // After Jetpack::load_modules()
+
+function jetpack_sitemap_load() {
+	if ( '1' == get_option( 'blog_public' ) ) { // loose comparison okay.
+		include_once 'sitemaps/sitemaps.php';
+	}
 }
 
 add_action( 'jetpack_activate_module_sitemaps', 'jetpack_sitemap_on_activate' );
```

Totally untested :)

I think b is better since it means the load order of the modules is irrelevant.

Reply:
I've tried a new Method 3 :). Please check [86d06ed](https://github.com/Automattic/jetpack/pull/10333/commits/86d06ed10e5ee1febdfc7a2332a21d35f1facf9c).

Reply:
I think this is fine for now, but we'll still probably want to make as-yet-unknown changes in the future around `blog_public`, how we want other developers to test for the privacy feature, etc.